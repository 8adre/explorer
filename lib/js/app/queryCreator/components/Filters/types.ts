import { Filter } from '../../types';

import {
  ADD_FILTER,
  UPDATE_FILTER,
  REMOVE_FILTER,
  RESET_FILTERS,
} from './constants';

interface AddFilterAction {
  type: typeof ADD_FILTER;
}

interface UpdateFilterAction {
  type: typeof UPDATE_FILTER,
  payload: {
    index: number;
    filter: Filter;
  };
}

interface RemoveFilterAction {
  type: typeof REMOVE_FILTER,
  payload: {
    index: number;
  };
}

interface ResetFiltersAction {
  type: typeof RESET_FILTERS
}

export type FiltersActions =
  | AddFilterAction
  | UpdateFilterAction
  | RemoveFilterAction
  | ResetFiltersAction;
