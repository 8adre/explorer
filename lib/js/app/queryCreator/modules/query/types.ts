import {
  SET_QUERY,
  SELECT_EVENT_COLLECTION,
  SELECT_ANALYSIS,
  SELECT_TARGET_PROPERTY,
  SET_PERCENTILE,
  SET_TIMEFRAME,
  SET_GROUP_BY,
  SET_ORDER_BY,
  SET_LIMIT,
  SELECT_TIMEZONE,
  ADD_FUNNEL_STEP,
  REMOVE_FUNNEL_STEP,
  UPDATE_FUNNEL_STEP,
  UPDATE_FUNNEL_STEP_EVENT_COLLECTION,
  CHANGE_FUNNEL_STEPS_ORDER,
} from './constants';

import { Timezones, Timeframe, OrderBy, FunnelStep } from '../../types';
import { Analysis } from '../../../types';

export type ReducerState = {
  eventCollection?: string;
  targetProperty?: string;
  percentile?: number;
  timezone?: number | Timezones;
  groupBy?: string | string[];
  orderBy?: string | OrderBy | OrderBy[];
  limit?: number;
  timeframe: Timeframe;
  analysisType: Analysis;
  steps: FunnelStep[];
};

export interface SetQuery {
  type: typeof SET_QUERY;
  payload: {
    query: Partial<ReducerState>;
  };
}

export interface SelectEventCollectionAction {
  type: typeof SELECT_EVENT_COLLECTION;
  payload: {
    name: string;
  };
}

export interface SelectAnalysisAction {
  type: typeof SELECT_ANALYSIS;
  payload: {
    analysis: Analysis;
  };
}

export interface SelectTargetPropertyAction {
  type: typeof SELECT_TARGET_PROPERTY;
  payload: {
    property: string;
  };
}

export interface SelectTimezoneAction {
  type: typeof SELECT_TIMEZONE;
  payload: {
    timezone: number | Timezones;
  };
}

export interface SetPercentileAction {
  type: typeof SET_PERCENTILE;
  payload: {
    percentile: number;
  };
}

export interface SetTimeframeAction {
  type: typeof SET_TIMEFRAME;
  payload: {
    timeframe: Timeframe;
  };
}

export interface SetGroupByAction {
  type: typeof SET_GROUP_BY;
  payload: {
    groupBy: string | string[] | undefined;
  };
}

export interface SetOrderByAction {
  type: typeof SET_ORDER_BY;
  payload: {
    orderBy: OrderBy[] | undefined;
  };
}

export interface SetLimitAction {
  type: typeof SET_LIMIT;
  payload: {
    limit: number | undefined;
  };
}

export interface AddFunnelStepAction {
  type: typeof ADD_FUNNEL_STEP;
}

export interface RemoveFunnelStepAction {
  type: typeof REMOVE_FUNNEL_STEP;
  payload: {
    index: number;
  };
}

export interface UpdateFunnelStepAction {
  type: typeof UPDATE_FUNNEL_STEP;
  payload: {
    index: number;
    properties: Partial<FunnelStep>;
  };
}

export interface UpdateFunnelStepEventCollectionAction {
  type: typeof UPDATE_FUNNEL_STEP_EVENT_COLLECTION;
  payload: {
    name: string;
  };
}

export interface ChangeFunnelStepsOrderAction {
  type: typeof CHANGE_FUNNEL_STEPS_ORDER;
  payload: {
    index: number;
    updatedIndex: number;
  };
}

export type QueryActions =
  | SetQuery
  | SelectEventCollectionAction
  | SelectAnalysisAction
  | SelectTargetPropertyAction
  | SetPercentileAction
  | SetGroupByAction
  | SetOrderByAction
  | SetLimitAction
  | SetTimeframeAction
  | SelectTimezoneAction
  | AddFunnelStepAction
  | UpdateFunnelStepAction
  | RemoveFunnelStepAction
  | UpdateFunnelStepEventCollectionAction
  | ChangeFunnelStepsOrderAction;
