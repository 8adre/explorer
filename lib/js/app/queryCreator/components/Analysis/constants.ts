import { AnalysisItem } from './types';

export const ANALYSIS_GROUPS: AnalysisItem[][] = [
  [
    {
      label: 'Average',
      value: 'average',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Count',
      value: 'count',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Count Unique',
      value: 'count_unique',
      description: 'lorem ipsum dolor sit amet',
    },
  ],
  [
    {
      label: 'Extraction',
      value: 'extraction',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Funnel',
      value: 'funnel',
      description: 'lorem ipsum dolor sit amet',
    },
  ],
  [
    {
      label: 'Maximum',
      value: 'maximum',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Median',
      value: 'median',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Minimum',
      value: 'minimum',
      description: 'lorem ipsum dolor sit amet',
    },
  ],
  [
    {
      label: 'Percentile',
      value: 'percentile',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Select unique',
      value: 'select_unique',
      description: 'lorem ipsum dolor sit amet',
    },
    {
      label: 'Standard deviation',
      value: 'standard_deviation',
      description: 'lorem ipsum dolor sit amet',
    },
    { label: 'Sum', value: 'sum', description: 'lorem ipsum dolor sit amet' },
  ],
];
