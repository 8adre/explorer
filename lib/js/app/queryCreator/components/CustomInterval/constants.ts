export const TIME_UNITS = [
  'minutes',
  'hours',
  'days',
  'weeks',
  'months',
  'years',
];

export const MINIMUM_VALUE = 1;
export const EVERY = 'every';
export const INTERVAL_SEPARATOR = '_';
