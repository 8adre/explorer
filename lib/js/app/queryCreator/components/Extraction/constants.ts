export const DEFAULT_LIMIT = 100;

export const GZIP_ENCODING = 'gzip';
