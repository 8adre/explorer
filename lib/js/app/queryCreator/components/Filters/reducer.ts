import {
  ADD_FILTER,
  UPDATE_FILTER,
  REMOVE_FILTER,
  RESET_FILTERS,
} from './constants';

import { FiltersActions } from './types';
import { Filter } from '../../types';

export const initialState: Filter[] = [];

type State = (Filter | null)[];

export const filtersReducer = (state: State, action: FiltersActions): State => {
  switch (action.type) {
    case RESET_FILTERS:
      return initialState;
    case ADD_FILTER:
      return [...state, null];
    case UPDATE_FILTER:
      return state.map((filter, idx) => {
        if (idx === action.payload.index) return action.payload.filter;
        return filter;
      });
    case REMOVE_FILTER:
      return state.filter((_, idx) => idx !== action.payload.index);
    default:
      return state;
  }
};
