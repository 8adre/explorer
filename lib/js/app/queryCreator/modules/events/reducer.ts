import { ReducerState, EventsActions } from './types';

import { createTree } from '../../utils/createTree';
import { createCollection } from '../../utils/createCollection';

import {
  SET_EVENTS_COLLECTIONS,
  FETCH_COLLECTION_SCHEMA_SUCCESS,
} from './constants';

export const initialState: ReducerState = {
  collections: [],
  schemas: {},
};

export const eventsReducer = (
  state: ReducerState = initialState,
  action: EventsActions
) => {
  switch (action.type) {
    case FETCH_COLLECTION_SCHEMA_SUCCESS:
      return {
        ...state,
        schemas: {
          ...state.schemas,
          [action.payload.collection]: {
            schema: action.payload.schema,
            tree: createTree(action.payload.schema),
            list: createCollection(action.payload.schema),
          },
        },
      };
    case SET_EVENTS_COLLECTIONS:
      return {
        ...state,
        collections: action.payload.collections,
      };
    default:
      return state;
  }
};
