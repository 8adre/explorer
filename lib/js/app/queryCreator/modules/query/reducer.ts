import {
  SELECT_EVENT_COLLECTION,
  SELECT_ANALYSIS,
  SELECT_TIMEZONE,
  SELECT_TARGET_PROPERTY,
  SET_QUERY,
  SET_PROPERTY_NAMES,
  SET_LIMIT,
  SET_EXTRACTION_LIMIT,
  SET_PERCENTILE,
  SET_TIMEFRAME,
  SET_GROUP_BY,
  SET_ORDER_BY,
  DEFAULT_ANALYSIS,
  DEFAULT_TIMEFRAME,
  ADD_FUNNEL_STEP,
  REMOVE_FUNNEL_STEP,
  UPDATE_FUNNEL_STEP,
  DEFAULT_FUNNEL_STEP,
} from './constants';

import { ReducerState, QueryActions } from './types';

export const initialState: ReducerState = {
  eventCollection: null,
  targetProperty: null,
  percentile: null,
  timezone: undefined,
  groupBy: undefined,
  orderBy: undefined,
  limit: undefined,
  timeframe: DEFAULT_TIMEFRAME,
  analysisType: DEFAULT_ANALYSIS,
  steps: [],
};

export const queryReducer = (
  state: ReducerState = initialState,
  action: QueryActions
) => {
  switch (action.type) {
    case SET_QUERY:
      return {
        ...initialState,
        ...action.payload.query,
      };
      case SET_LIMIT:
      return {
        ...state,
        limit: action.payload.limit,
      };
      case SET_EXTRACTION_LIMIT:
      return {
        ...state,
        latest: action.payload.limit,
      };
    case SET_PROPERTY_NAMES:
      return {
        ...state,
        propertyNames: action.payload.propertyNames,
      };
    case UPDATE_FUNNEL_STEP:
      return {
        ...state,
        steps: state.steps.map((step, index) => {
          if (index === action.payload.index) {
            return {
              ...step,
              ...action.payload.properties,
            };
          }
          return step;
        }),
      };
    case REMOVE_FUNNEL_STEP:
      return {
        ...state,
        steps: state.steps.filter(
          (_step, index) => index !== action.payload.index
        ),
      };
    case ADD_FUNNEL_STEP:
      return {
        ...state,
        steps: [...state.steps, DEFAULT_FUNNEL_STEP],
      };
    case SET_ORDER_BY:
      return {
        ...state,
        orderBy: action.payload.orderBy,
      };
    case SET_GROUP_BY:
      return {
        ...state,
        groupBy: action.payload.groupBy,
      };
    case SET_TIMEFRAME:
      return {
        ...state,
        timeframe: action.payload.timeframe,
      };
    case SET_PERCENTILE:
      return {
        ...state,
        percentile: action.payload.percentile,
      };
    case SELECT_TIMEZONE:
      return {
        ...state,
        timezone: action.payload.timezone,
      };
    case SELECT_TARGET_PROPERTY:
      return {
        ...state,
        targetProperty: action.payload.property,
      };
    case SELECT_ANALYSIS:
      return {
        ...state,
        analysisType: action.payload.analysis,
      };
    case SELECT_EVENT_COLLECTION:
      return {
        ...state,
        eventCollection: action.payload.name,
      };
    default:
      return state;
  }
};
