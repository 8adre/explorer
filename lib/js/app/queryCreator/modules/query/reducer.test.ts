import { queryReducer, initialState } from './reducer';

import {
  resetExtraction,
  setQuery,
  setLimit,
  setTimeframe,
  setExtractionLimit,
  setExtractionRecipientEmail,
  setExtractionContentEncoding,
  setPropertyNames,
  updateFunnelStep,
  removeFunnelStep,
  addFunnelStep,
  setGroupBy,
  setOrderBy,
  setInterval,
  setFilters,
  selectAnalysis,
  selectTargetProperty,
  selectEventCollection,
  setPercentile,
} from './actions';
import { DEFAULT_FUNNEL_STEP } from './constants';

import { ReducerState } from './types';
import { OrderBy, Filter } from '../../types';

test('reset extraction properties to initial state', () => {
  const action = resetExtraction();
  const state = queryReducer(
    {
      ...initialState,
      contentEncoding: 'gzip',
      propertyNames: ['surname'],
    },
    action
  );

  expect(state).toEqual(initialState);
});

test('set query properties', () => {
  const properties = {
    analysisType: 'extraction',
    groupBy: ['country'],
  };
  const action = setQuery(properties as Partial<ReducerState>);
  const state = queryReducer(initialState, action);

  expect(state).toEqual({
    ...initialState,
    ...properties,
  });
});

test('set query limit', () => {
  const action = setLimit(100);
  const { limit } = queryReducer(initialState, action);

  expect(limit).toEqual(100);
});

test('set extraction limit', () => {
  const action = setExtractionLimit(45);
  const { latest } = queryReducer(initialState, action);

  expect(latest).toEqual(45);
});

test('set extraction email recipient', () => {
  const email = 'user@keen.io';
  const action = setExtractionRecipientEmail(email);
  const state = queryReducer(initialState, action);

  expect(state.email).toEqual(email);
});

test('set extraction content encoding', () => {
  const encoding = 'gzip';
  const action = setExtractionContentEncoding(encoding);

  const state = queryReducer(initialState, action);

  expect(state.contentEncoding).toEqual(encoding);
});

test('set extraction property names', () => {
  const properties = ['name', 'surname', 'id'];
  const action = setPropertyNames(properties);

  const state = queryReducer(initialState, action);

  expect(state.propertyNames).toEqual(properties);
});

test('set query timeframe', () => {
  const timeframe = 'this_100_days';
  const action = setTimeframe(timeframe);
  const state = queryReducer(initialState, action);

  expect(state.timeframe).toEqual(timeframe);
});

test('set query order by properties', () => {
  const orderBy: OrderBy[] = [
    {
      propertyName: 'result',
      direction: 'ASC',
    },
  ];

  const action = setOrderBy(orderBy);
  const state = queryReducer(initialState, action);

  expect(state.orderBy).toEqual(orderBy);
});

test('set query interval', () => {
  const interval = 'monthly';
  const action = setInterval(interval);
  const state = queryReducer(initialState, action);

  expect(state.interval).toEqual(interval);
});

test('set group by properties', () => {
  const groupBy = ['name', 'surname'];

  const action = setGroupBy(groupBy);
  const state = queryReducer(initialState, action);

  expect(state.groupBy).toEqual(groupBy);
});

test('set percentile', () => {
  const percentile = 85;

  const action = setPercentile(percentile);
  const state = queryReducer(initialState, action);

  expect(state.percentile).toEqual(percentile);
});

test('set analysis', () => {
  const analysis = 'funnel';

  const action = selectAnalysis(analysis);
  const state = queryReducer(initialState, action);

  expect(state.analysisType).toEqual(analysis);
});

test('set target property', () => {
  const property = 'keen.id';

  const action = selectTargetProperty(property);
  const state = queryReducer(initialState, action);

  expect(state.targetProperty).toEqual(property);
});

test('set event collection', () => {
  const collection = 'logins';

  const action = selectEventCollection(collection);
  const state = queryReducer(initialState, action);

  expect(state.eventCollection).toEqual(collection);
});

test('set filters', () => {
  const filters = [{"propertyName":"state","operator":"ne","propertyValue":"Arizona"}] as Filter[];

  const action = setFilters(filters);
  const state = queryReducer(initialState, action);

  expect(state.filters).toEqual(filters);
});

test('update funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
      },
    ],
  };

  const action = updateFunnelStep(0, {
    withActors: true,
    eventCollection: 'logins',
  });

  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});

test('removes funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
      },
    ],
  };

  const action = removeFunnelStep(0);
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toEqual([]);
});

test('add funnel step', () => {
  const stepsState = {
    steps: [
      {
        ...DEFAULT_FUNNEL_STEP,
        eventCollection: 'clicks',
      },
    ],
  };

  const action = addFunnelStep();
  const { steps } = queryReducer(
    {
      ...initialState,
      ...stepsState,
    },
    action
  );

  expect(steps).toMatchSnapshot();
});
