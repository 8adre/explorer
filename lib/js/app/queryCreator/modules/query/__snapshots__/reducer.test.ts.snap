// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add funnel step 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "clicks",
    "filters": Array [],
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
  Object {
    "actorProperty": undefined,
    "eventCollection": undefined,
    "filters": Array [],
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": false,
  },
]
`;

exports[`update funnel step 1`] = `
Array [
  Object {
    "actorProperty": undefined,
    "eventCollection": "logins",
    "filters": Array [],
    "inverted": false,
    "optional": false,
    "timeframe": "this_14_days",
    "timezone": undefined,
    "withActors": true,
  },
]
`;
